sw/device/lib/dif/dif_otp_ctrl.h

```diff
@@ -0,0 +1,944 @@
+  /**
+   * Indicates that the hardware initialization sequence failed.
+   *
+   * This error indicates non-recoverable hardware malfunction.
+   */
+  kDifOtpCtrlErrorFailedInit,
+  /**
+   * Indicates a recoverable error during a read operation.
+   *
+   * A followup read should work as expected.
+   */
+  kDifOtpCtrlErrorRecoverableRead,
+  /**
+   * Indicates an unrecoverable error during a read operation.
+   *
+   * This error indicates non-recoverable hardware malfunction.
+   */
+  kDifOtpCtrlErrorUnrecoverableRead,
+  /**
+   * Indicates an unspecified error during a read operation.
+   *
+   * This error indicates non-recoverable hardware malfunction.
+   */
+  kDifOtpCtrlErrorUnspecifiedRead,
+  /**
+   * Indicates that the blank write check failed during a write operation.
+   */
+  kDifOtpCtrlErrorBlankCheckFailed,
+  /**
+   * Indicates an unspecified error during a write operation.
+   *
+   * This error indicates non-recoverable hardware malfunction.
+   */
+  kDifOtpCtrlErrorUnspecifiedWrite,
+  /**
+   * Indicates an invalid command was issue to the DAI.
+   */
+  kDifOtpCtrlErrorInvalidDaiCommand,
+  /**
+   * Indicates a locked memory region was accessed.
+   */
+  kDifOtpCtrlErrorLockedAccess,
+  /**
+   * Indicates an integrity check failed in the buffer registers.
+   *
+   * This error indicates non-recoverable hardware malfunction.
+   */
+  kDifOtpCtrlErrorIntgrityLoss,
+  /**
+   * Indicates an consistency check failed in the buffer registers.
+   *
+   * This error indicates non-recoverable hardware malfunction.
+   */
+  kDifOtpCtrlErrorConsistencyLoss,
+  /**
+   * Indicates that the FSM of the controller is in a bad state.
+   *
+   * This error indicates that the device has been glitched by an attacker.
+   */
+  kDifOtpCtrlErrorFsmBadState,
+  /**
+   * Indicates that an escalation occured in the alert subsystem, and the
+   * controller's FSM is in its terminal state.
+   */
+  kDifOtpCtrlErrorFsmTerminalState,
+} dif_otp_ctrl_error_t;
+
+/**
+ * The overall status of the OTP controller.
+ *
+ * See `dif_otp_ctrl_get_status()`.
+ */
+typedef struct dif_otp_ctrl_status {
+  /**
+   * Currently active statuses, given as a bit vector. To check whether a
+   * particular status code was returned, write
+   *
+   *   bool has_code = (status.codes >> kMyStatusCode) & 1;
+   *
+   * Note that it is possible to quickly check that the controller is idle and
+   * error-free by writing
+   *
+   *   bool is_ok = status.codes == (1 << kDifOtpStatusCodeDaiIdle);
+   */
+  uint32_t codes;
+  /**
+   * A list of root causes for each error status code.
+   *
+   * If the error status code `error` is present in `codes`, then
+   * `causes[error]` will contain its root cause.
+   */
+  // Open Q: This array is much wider than it needs to be, since not all status
+  // codes have associated errors. We could make this array shorter, since the
+  // statuses that *can* have errors are the prefix of the enum. Thoughts?
```

Comment:
@lenary @gkelly How about we introduce a constant to the effect of "any enum value <= to this one can be root-caused" and use that as the length of the array?

Reply:
That sounds reasonable. I only want to make sure we're not doing arithmetic on these values to build enums.

Reply:
A dedicated function could be an overkill and it looks like callers would have to access `codes` and `causes` multiple times to get more information anyway, so how about
- moving the first 9 constants in `dif_otp_ctrl_status_code_t` to a separate enum,
- defining a single `k...ErrorInModules` (tech spec calls these _modules_) in `dif_otp_ctrl_status_code_t` constants, and
- defining the array to be a 9 element array as you said?

Then, clients need to check this array using the new enums if `k...ErrorInModules` is set in `codes`. This can be slightly safer than just a comment.

But, IDK how LCI and DAI can have the same error codes as partitions while they are not partitions themselves.

Reply:
The LCI, DAI, and the partition controllers share some of the error codes, since they talk to the same physical memory macro underneath via an arbitrated bus. 

So in order to avoid having to remap the same errors to different encodings, I defined just one single enum with the super set of possible codes.

For reference, https://github.com/lowRISC/opentitan/pull/3456 adds a table that lists all possible error codes and where they can occur (look in the "Error Handling" section). I think in that specific context, I named these interfaces and partitions just "agents". Let me know if you think we should name them differently....

Reply:
I don't think the separate enum helps; you still need to know which ones are actually usable for indexing into the array. I think having a constant is the least-bad option.

Reply:
All constants in that enum would be usable for indexing into the array but it's up to you.

Reply:
I wound up going with the extra enum constant. it's not amazing, but C isn't exactly going to save us from ourselves in either case...